lib/workers/repository/process/vulnerabilities.ts

```diff
@@ -91,51 +97,392 @@ export class Vulnerabilities {
+        );
+        return [];
+      }
+
+      for (const vulnerability of vulnerabilities) {
+        for (const affected of vulnerability.affected ?? []) {
+          const isVulnerable = this.isPackageVulnerable(
+            ecosystem,
+            packageName,
+            depVersion,
+            affected,
+            versioningApi
+          );
+          if (!isVulnerable) {
+            continue;
+          }
+
+          logger.debug(
+            `Vulnerability ${vulnerability.id} affects ${packageName} ${depVersion}`
+          );
+          const fixedVersion = this.getFixedVersion(
+            ecosystem,
+            depVersion,
+            affected,
+            versioningApi
+          );
+          if (is.nullOrUndefined(fixedVersion)) {
+            logger.debug(
+              `No fixed version available for vulnerability ${vulnerability.id} in ${packageName} ${depVersion}`
+            );
+            continue;
+          }
+
+          logger.debug(
+            `Setting allowed version ${fixedVersion} to fix vulnerability ${vulnerability.id} in ${packageName} ${depVersion}`
+          );
+          const rule = this.convertToPackageRule(
+            packageFileConfig,
+            dep,
+            packageName,
+            depVersion,
+            fixedVersion,
+            vulnerability
+          );
+          packageRules.push(rule);
+        }
+      }
+
+      this.sortByFixedVersion(packageRules, versioningApi);
+    } catch (err) {
+      logger.debug(
+        { err },
+        `Error fetching vulnerability information for ${packageName}`
+      );
+    }
+
+    return packageRules;
+  }
+
+  private sortByFixedVersion(
+    packageRules: PackageRule[],
+    versioningApi: VersioningApi
+  ): void {
+    const versionsCleaned: Record<string, string> = {};
+    for (const rule of packageRules) {
+      const version = rule.allowedVersions as string;
+      versionsCleaned[version] = version.replace(regEx(/[=> ]+/g), '');
+    }
 
-    const vulnerabilities = await this.osvOffline?.getVulnerabilities(
-      ecosystem!,
-      packageDependency.depName!
+    packageRules.sort((a, b) =>
+      versioningApi.sortVersions(
+        versionsCleaned[a.allowedVersions as string],
+        versionsCleaned[b.allowedVersions as string]
+      )
     );
-    return this.convertToPackageRule(
-      vulnerabilities ?? [],
-      packageDependency.depName!,
-      ecosystem!
+  }
+
+  // https://ossf.github.io/osv-schema/#affectedrangesevents-fields
+  private sortEvents(
+    events: Osv.Event[],
+    versioningApi: VersioningApi
+  ): Osv.Event[] {
+    const sortedCopy: Osv.Event[] = [];
+    let zeroEvent: Osv.Event | null = null;
+
+    for (const event of events) {
+      if (event.introduced === '0') {
+        zeroEvent = event;
+        continue;
+      }
+      sortedCopy.push(event);
+    }
+
+    sortedCopy.sort((a, b) =>
+      versioningApi.sortVersions(Object.values(a)[0], Object.values(b)[0])
```

Comment:
we then should make a record cache again, because those calls are very expensive.

Reply:
In general, true. I ran a check over all 13937 advisories and found that on average this operation would sort only **4 values** (3.938 to be precise, a 0 version wouldn't be sorted).

Changing this would most probably require building a map based on object indices, since the keys can occur multiple times. In the sort operation itself, I'd then have no access to these indices, as `a` and `b` don't have this info.

So overall, I think that in this specific case the gain is simply not worth the effort and extra code.

Reply:
ok, please add a comment to the code about that.